### DISCUSSION TITLE
Error found: HTTP Error 403: rate limit exceeded

### DISCUSSION THREAD

**Discussion by ravikri-bh** [thoughtspot/cs_tools#111] [General]
I am working on cs_tools installation on an EC2 instance running with Amazon Linux 2023.

I see the below error log, need inputs/assistance on resolving the error.

> INFO     | 2024-03-13 14:13:34,873 | <stdin>:92 | Welcome to the CS Tools Bootstrapper!
>          | Ideally, you will only need to run this one time, and then your environment can be fully managed by CS Tools itself.
>          | If you run into any issues, please reach out to us on GitHub Discussions below.
>          |
>          |           GitHub: https://github.com/thoughtspot/cs_tools/issues/new/choose
>          |
> DEBUG    | 2024-03-13 14:13:34,878 | <stdin>:107 |
>     [PLATFORM DETAILS]
>     Python Version: 3.9.16
>        System Info: Linux (detail: Linux-6.1.72-96.166.amzn2023.x86_64-x86_64-with-glibc2.34)
>      Platform Tags: linux-x86_64
>             Ran at: 2024-03-13 14:13:34
> 
> INFO     | 2024-03-13 14:13:34,878 | <stdin>:395 | Missing '_updater.py', downloading from GitHub
> DEBUG    | 2024-03-13 14:13:34,953 | <stdin>:519 | Error found: HTTP Error 403: rate limit exceeded
> Traceback (most recent call last):
>   File "<stdin>", line 514, in main
>   File "<stdin>", line 123, in cli
>   File "<stdin>", line 397, in get_cs_tools_venv
>   File "<stdin>", line 481, in http_request
>   File "/usr/lib64/python3.9/urllib/request.py", line 214, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib64/python3.9/urllib/request.py", line 523, in open
>     response = meth(req, response)
>   File "/usr/lib64/python3.9/urllib/request.py", line 632, in http_response
>     response = self.parent.error(
>   File "/usr/lib64/python3.9/urllib/request.py", line 561, in error
>     return self._call_chain(*args)
>   File "/usr/lib64/python3.9/urllib/request.py", line 494, in _call_chain
>     result = func(*args)
>   File "/usr/lib64/python3.9/urllib/request.py", line 641, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> urllib.error.HTTPError: HTTP Error 403: rate limit exceeded
> 

Thank you

**Comment by boonhapus** [comment]
@ravikri-bh I've never seen this before. We only make a single request over to GitHub, so I'm unsure how you reached this error 🤔

Is this happening consistently when trying to install?

**Reply by ravikri-bh** [reply]
@boonhapus Yes, I see the same error log every time I attempt to install on an EC2.

**Reply by boonhapus** [reply]
@ravikri-bh there's no way you've made 5,000 requests in an hour... haha let me check some things internally and get back to you. We might need to set up time directly to debug this one.

Can you share the output of `python --m sysconfig` ?

**Reply by ravikri-bh** [reply]
@boonhapus , that's right. I tried installing 2 times ending with same error log. The EC2 instance was created yesterday for running cs_tools exclusively.

As requested please see the output of `python3 -m sysconfig` below

> $ python3 -m sysconfig
> Platform: "linux-x86_64"
> Python version: "3.9"
> Current installation scheme: "posix_prefix"
> 
> Paths:
>         data = "/usr"
>         include = "/usr/include/python3.9"
>         platinclude = "/usr/include/python3.9"
>         platlib = "/usr/lib64/python3.9/site-packages"
>         platstdlib = "/usr/lib64/python3.9"
>         purelib = "/usr/lib/python3.9/site-packages"
>         scripts = "/usr/bin"
>         stdlib = "/usr/lib64/python3.9"

**Reply by boonhapus** [reply]
Thanks @ravikri-bh , I will build you an offline installer to get around this issue.

**Reply by ravikri-bh** [reply]
@boonhapus , please share the link to download the offline installer.

**Reply by boonhapus** [reply]
Hi @ravikri-bh , please find it below.

Can you try the steps below?

1. Extract the zip file.
2. Run python against the `_bootstrapper.py` file with `--offline-mode` specified

`cd cs-tools_1.4.16_linux-x86_64_3.9`
`python _bootstrapper.py --install --offline-mode .`

It's a bit too big for GitHub discussions, so I've attached an Egnyte link. Let me know if that works for you!
**[https://thoughtspot.egnyte.com/dl/RdF3OIxjBK/cs-tools_v1.4.16_linux-x86_64_3.9.zip](https://thoughtspot.egnyte.com/dl/2ms5grPoir/cs-tools_v1.4.16_linux-x86_64_3.9.zip_)**
